File: solutions/cpp/clock/1/clock.cpp
#include "clock.h"
#include <iomanip>
#include <sstream>

namespace date_independent {

clock clock::at(int hour, int minute) {
    constexpr int MINUTES_PER_HOUR = 60;
    constexpr int HOURS_PER_DAY = 24;
    hour = (hour + minute / MINUTES_PER_HOUR) % HOURS_PER_DAY;
    minute = minute % MINUTES_PER_HOUR;
    if (minute < 0) {
        hour--;
        minute += MINUTES_PER_HOUR;
    }
    if (hour < 0) hour += HOURS_PER_DAY;
    return {hour, minute};
}
    
clock clock::plus(int minutes) {
    return at(hour, this->minute + minutes);
}

static std::string format(int n) {
    std::ostringstream result;
    result << std::setw(2) << std::setfill('0') << n;
    return result.str();
}

clock::operator std::string() const {
    return format(hour) + ":" + format(minute);
}
    
bool clock::operator ==(const clock& other) const {
    return this->hour == other.hour && this->minute == other.minute;
}

bool clock::operator !=(const clock& other) const {
    return !(*this == other);
}

}  // namespace date_independent

File: solutions/typescript/clock/1/clock.ts
export class Clock {

  minutes: number;
  readonly MINUTES_PER_DAY = 24 * 60;
  
  constructor(hour: number, minute?: number) {
    let minutes = (hour * 60 + (minute || 0)) % this.MINUTES_PER_DAY;
    let rollover = minutes < 0 ? this.MINUTES_PER_DAY : 0;
    this.minutes = minutes + rollover;
  }

  public toString(): string {
    let hour = Math.trunc(this.minutes / 60);
    let minute = this.minutes % 60;
    return `${this.pad(hour)}:${this.pad(minute)}`;
  }

  pad(n: number): string {
    return n.toString().padStart(2, '0');
  }

  public plus(minutes: number): Clock {
    return new Clock(0, this.minutes + minutes);
  }

  public minus(minutes: number): Clock {
    return new Clock(0, this.minutes - minutes);
  }

  public equals(other: Clock): boolean {
    return this.minutes == other.minutes;
  }
}
